Adding Docker-specific instructions to upgrade Elasticsearch (#4366)
elastic/docs-content: deploy-manage/upgrade/deployment-or-cluster/upgrade-elasticsearch-docker.md

---
navigation_title: Upgrade Elasticsearch running on Docker
applies_to:
  deployment:
    self: ga
products:
  - id: elasticsearch
---

# Upgrading the {{es}} version running on Docker

> You update {{es}} running in a Docker container by pulling the new Docker image and restarting the container with the new image.

Reviewer: ```suggestion
You update {{es}} running in a Docker container by pulling the new Docker image and recreating the container using the new image and the same data volumes as the original container.
```

Author: Thanks, Edu. I've updated this in a separate commit

Replace `<x.y.z>` with the version of the Docker image you downloaded.

    ```sh subs=true
    wget https://artifacts.elastic.co/cosign.pub
    cosign verify --key cosign.pub docker.elastic.co/elasticsearch/elasticsearch:<x.y.z>
    ```

    The `cosign` command prints the check results and the signature payload in JSON format:

    ```sh subs=true
    Verification for docker.elastic.co/elasticsearch/elasticsearch:<x.y.z> --
    The following checks were performed on each of these signatures:
      - The cosign claims were validated
      - Existence of the claims in the transparency log was verified offline
      - The signatures were verified against the specified public key
    ```

1. Stop the currently running {{es}} container. Replace `<container_name>` with the name or ID of your {{es}} container.

    ```shell
    docker stop <container_name>
    ```

1. After the container has stopped, remove it. Correctly mapping your data directory to a volume outside of the container ensures that your data is not deleted. Replace `<container_name>` with the name or ID of your {{es}} container.

    ```shell
    docker rm <container_name>

>     ```

Author: I've added an important admonition that specifies our assumptions around this example. Wdyt?

* If you're running a cluster, you'll need to carefully plan the upgrade process to minimize downtime and ensure cluster stability. This often involves a rolling upgrade strategy.
* When using a `docker-compose.yml` make sure to update the desired version in the configuration file or the environment variable.

## Hardened Docker images [docker-wolfi-images]

You can also use the hardened [Wolfi](https://wolfi.dev/) image for additional security. Using Wolfi images requires Docker version 20.10.10 or higher.

To use the Wolfi image, append `-wolfi` to the image tag in the Docker command.

For example:

```sh subs=true
docker pull docker.elastic.co/elasticsearch/elasticsearch-wolfi:<x.y.z>
```

Replace `<x.y.z>` with the version number you want to upgrade to.

## Upgrade {{es}} running on Docker [upgrade-process]

1. Pull the new version of the {{es}} Docker image from Elastic's Docker registry using the `docker pull` command. Replace `<x.y.z>` with the version number you want to upgrade to.

    ```shell
    docker pull docker.elastic.co/elasticsearch/elasticsearch:<x.y.z>
    ```

> 1. Optional: Install [Cosign](https://docs.sigstore.dev/cosign/system_config/installation/) for your environment. Then use Cosign to verify the {{es}} image’s signature.

Author: Edu's explanation is on point -- I use this syntax because it's the easiest to maintain (and allows you to future-proof further edits). It renders as an ordered list. 

Here is the [preview for this page](https://docs-v3-preview.elastic.dev/elastic/docs-content/pull/4366/deploy-manage/upgrade/deployment-or-cluster/upgrade-elasticsearch-docker)

Replace `<x.y.z>` with the version of the Docker image you downloaded.

    ```sh subs=true
    wget https://artifacts.elastic.co/cosign.pub
    cosign verify --key cosign.pub docker.elastic.co/elasticsearch/elasticsearch:<x.y.z>
    ```

    The `cosign` command prints the check results and the signature payload in JSON format:

    ```sh subs=true
    Verification for docker.elastic.co/elasticsearch/elasticsearch:<x.y.z> --
    The following checks were performed on each of these signatures:
      - The cosign claims were validated
      - Existence of the claims in the transparency log was verified offline
      - The signatures were verified against the specified public key
    ```

1. Stop the currently running {{es}} container. Replace `<container_name>` with the name or ID of your {{es}} container.

    ```shell
    docker stop <container_name>
    ```

> 1. After the container has stopped, you can remove it. Correctly mapping your data directory to a volume outside of the container ensures that your data is not deleted. Replace `<container_name>` with the name or ID of your {{es}} container.

Reviewer: ```suggestion
1. After the container has stopped, you can remove it. If your data directory is correctly mapped to a volume outside of the container, your data is preserved. Replace `<container_name>` with the name or ID of your {{es}} container.
```

    :::{important}
    This example assumes that you've started the original container with the configuration and data directories stored in separate volumes.
    :::

    ```shell
    docker rm <container_name>
    ```

> 1. Start a new container using the new image. Use the same volume mappings and configuration settings as the old container to ensure that your data and configuration are preserved. Replace `<container_name>` with the name you want for your new container.

Author: That's a good point, @jozala! Thank you for spotting all these inconsistencies.

I've addressed your suggestions in [9a19c1f](9a19c1f7fff70bd141239077e38d787441d8afd8).
Hopefully the doc looks better now, wdyt?

    ```shell
    docker run --name <container_name> -p 9200:9200 -p 9300:9300 \
    -e "discovery.type=single-node" \
    -v path_to_data_volume:/usr/share/elasticsearch/data \
    -v path_to_config_volume:/usr/share/elasticsearch/config \
    docker.elastic.co/elasticsearch/elasticsearch:<x.y.z>
    ```

    Adjust the `-p` flags for port mappings, `-e` for environment variables, and `-v` for volume mappings as needed based on your setup.

1. After the new container starts, verify that {{es}} is running the new version by querying the root URL of your {{es}} instance. 

    ```shell
    curl http://localhost:9200

>     docker.elastic.co/elasticsearch/elasticsearch:<x.y.z>

Reviewer: ```suggestion
    curl http://localhost:9200
```

Author: I missed this! Thanks, Edu
